src/storage/v2/inmemory/edge_type_property_index.cpp

```diff
@@ -0,0 +1,332 @@
+  ReturnType link;
+  Delta *delta = nullptr;
+  {
+    auto guard = std::shared_lock{vertex->lock};
+    delta = vertex->delta;
+  }
+  ApplyDeltasForRead(transaction, delta, view, [&edge, &vertex, &link](const Delta &delta) {
+    switch (delta.action) {
+      case Delta::Action::ADD_LABEL:
+      case Delta::Action::REMOVE_LABEL:
+      case Delta::Action::SET_PROPERTY:
+        break;
+      case Delta::Action::ADD_IN_EDGE: {
+        if (edge == delta.vertex_edge.edge.ptr) {
+          link = {delta.vertex_edge.edge_type, delta.vertex_edge.vertex, delta.vertex_edge.edge};
+          auto it = std::find(vertex->in_edges.begin(), vertex->in_edges.end(), link);
+          MG_ASSERT(it == vertex->in_edges.end(), "Invalid database state!");
+          break;
+        }
+      }
+      case Delta::Action::ADD_OUT_EDGE: {
+        if (edge == delta.vertex_edge.edge.ptr) {
+          link = {delta.vertex_edge.edge_type, delta.vertex_edge.vertex, delta.vertex_edge.edge};
+          auto it = std::find(vertex->out_edges.begin(), vertex->out_edges.end(), link);
+          MG_ASSERT(it == vertex->out_edges.end(), "Invalid database state!");
+          break;
+        }
+      }
+      case Delta::Action::REMOVE_IN_EDGE:
+      case Delta::Action::REMOVE_OUT_EDGE:
+      case Delta::Action::RECREATE_OBJECT:
+      case Delta::Action::DELETE_DESERIALIZED_OBJECT:
+      case Delta::Action::DELETE_OBJECT:
+        break;
+    }
+  });
+  return link;
+}
+}  // namespace
+
+namespace memgraph::storage {
+
+bool InMemoryEdgeTypePropertyIndex::Entry::operator<(const Entry &rhs) const {
+  if (edge->gid < rhs.edge->gid) {
+    return true;
+  }
+  if (rhs.edge->gid < edge->gid) {
+    return false;
+  }
+  return std::make_tuple(value, from_vertex, to_vertex, edge, timestamp) <
+         std::make_tuple(rhs.value, rhs.from_vertex, rhs.to_vertex, rhs.edge, rhs.timestamp);
+}
+
+bool InMemoryEdgeTypePropertyIndex::Entry::operator==(const Entry &rhs) const {
+  return std::make_tuple(value, from_vertex, to_vertex, edge, timestamp) ==
+         std::make_tuple(rhs.value, rhs.from_vertex, rhs.to_vertex, rhs.edge, rhs.timestamp);
+}
+
+bool InMemoryEdgeTypePropertyIndex::Entry::operator<(const PropertyValue &rhs) const { return value < rhs; }
+
+bool InMemoryEdgeTypePropertyIndex::Entry::operator==(const PropertyValue &rhs) const { return value == rhs; }
+
+bool InMemoryEdgeTypePropertyIndex::CreateIndex(EdgeTypeId edge_type, PropertyId property,
+                                                utils::SkipList<Vertex>::Accessor vertices) {
+  auto [it, emplaced] = index_.try_emplace({edge_type, property});
+  if (!emplaced) {
+    return false;
+  }
+
+  utils::MemoryTracker::OutOfMemoryExceptionEnabler oom_exception;
+  try {
+    auto edge_acc = it->second.access();
+    for (auto &from_vertex : vertices) {
+      if (from_vertex.deleted) {
+        continue;
+      }
```

Comment:
During indexing there can be no other operation in database, that holds still, right?

Reply:
Yes, we take a unique lock in https://github.com/memgraph/memgraph/blob/impl-edge-type-property-indices/src/storage/v2/inmemory/storage.cpp#L1467